gitops/rings/README.md

```diff
@@ -0,0 +1,148 @@
+# Rings
+
+Ring deployment is a configuration on top of a service deployment that allows you to deploy *revisions* of the service alongside any existing instances of *that* service, and any other services. It is a variation of canary deployments where new feature releases are gradually deployed to production without the risk of affecting all end users.
+
+This README serves to explain a Ring based deployment using Fabrikate and Bedrock without using a Service Mesh.
+
+The Ring workflow is shown in the following diagram, where you will see that it represents an extension to the [Bedrock CI/CD](https://github.com/microsoft/bedrock/tree/master/gitops).
+
+![Ring Workflow](./images/ring-workflow.png)
+
+In summary, there are two additions made to the Bedrock CI/CD to account for this Rings implementation:
+
+1. Cluster HLD Repo
+2. Ring.yaml
+
+## Components of the Ringed Model
+
+### Git Repositories
+
+Recall that in the official Bedrock CI/CD (without Rings), there exists three repositories: (1) Service Source Code (2) Service HLD and (3) the Materialized Manifest. The concept of Rings introduces a *new* repository to the workflow, the Cluster HLD. Altogether, the following exists in the Rings workflow:
+
+For every independent service we assume 2 git repositories exist:
+
+**Service Source Repository**: A git repository that maintains the source code of the service, a dockerfile, and a helm chart. Developers will commit regularly to this repository, with revisions and rings being tracked in Git branches.
+
+**Service HLD Repository**: A second git repository that maintains a High Level Definition for the source repository. Commits to this repository are automated, and configuration is performed via AzDo pipelines. Subcomponents in this repository map to the branches in the Service Source Repository
+
+For all services represented by the above 2 git repositories, we assume two more repositories exist:
+
+**Cluster HLD Repository**: A git repository that maintains a High Level Definition for all Services and Revisions that are intended to be run on the Cluster.
+
+**“Materialized” Manifest Repository**: this git repository acts as our canonical source of truth for Flux – the in-cluster component that pulls and applies Kubernetes manifests rendered from the Cluster HLD repository.
+
+### Ring.yaml
+As a ring is considered to be strictly a revision of a microservice, we need a way to configure the ingress controller to route to the microservice revision a user belongs to. We achieve this by providing a `ring.yaml` file in our helm chart, which is an abstraction on Kubernetes and Traefik primitives.
+
+An example of ring.yaml:
+
+```yaml
```

Comment:
This file is actually the ring.yaml that gets output via the helm chart - not the file that gets consumed.

The file that we want for configuration is actually this:

```
serviceName: "languages"
majorVersion: "v1"
ringName: "48883-add-search"
group: "CMS-CORE"
contact: cms-core-dl@contoso.com
branch: feature/48883-add-search
initialUsers:
  - bill@contoso.com
  - ted@contoso.com
deploy: true

application:
  replicaCount: 10
  ports:
    - 8080
```